### PR TITLE
Added initial implementation of global 'single-pass' statistics

### PR DESCRIPTION
**Pull request by astrofrog**
This is a work in progress but the idea is to address #659 by adding a ``statistics`` method which finds stats over the whole cube (so no ``axis=`` keyword) using a single pass over chunks. This means a chunk is read, and all statistics are evaluated on it before moving to the next chunk. This can mean minimizing chunk I/O.

```python
In [1]: from spectral_cube import SpectralCube                                                                                                         

In [2]: cube = SpectralCube.read('l1448_13co_small.fits', use_dask=True)                                                                               

In [3]: cube.statistics()                                                                                                                              
Out[3]: 
{'min': <Quantity -0.66045946>,
 'max': <Quantity 4.0023365>,
 'sum': <Quantity 303203.88>,
 'mean': <Quantity 0.51889595>,
 'std': <Quantity 0.57547368>}
```

This is still incomplete, in particular it needs:

* [x] Changelog entry
* [x] Tests
* [x] Other 'easy' stats, e.g. count (of non-masked items) and rms
* [x] Documention
* [ ] A more robust implementation for std
* [ ] An equivalent method on the non-dask class? (not sure about this since it wouldn't be 'efficient' in the same way as for dask)

Maybe in a separate PR we can then work on adding optional 'robust' statistics (like for CASA's ``ia.statistics``) but this requires first implementing algorithms for parallel percentiles (I have ideas about this).

In the mean time, @e-koch and @keflavich, maybe you could test this out to see if the performance is reasonable (and how it compares to CASA).

### CODE REVIEW

**Comment by e-koch** [comment]
@astrofrog -- I'm getting the following error. Might be related to the block sizes, but they're all the same size (i.e. no rounding up/down # of pix) so I'm unsure where this is coming from.

```
In [12]: out = cube.statistics()                                                                                   
[########################################] | 100% Completed |  7min  1.1s
[########################################] | 100% Completed |  7min  1.2s
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
<ipython-input-12-4cf117b4bb9b> in <module>
----> 1 out = cube.statistics()

~/ownCloud/code_development/radio_astro_tools/spectral-cube/spectral_cube/dask_spectral_cube.py in wrapper(self, *args, **kwargs)
     64         with warnings.catch_warnings():
     65             warnings.simplefilter('ignore')
---> 66             return function(self, *args, **kwargs)
     67 
     68     return wrapper

~/ownCloud/code_development/radio_astro_tools/spectral-cube/spectral_cube/dask_spectral_cube.py in statistics(self)
    726                              np.nansum(chunk * chunk)])
    727 
--> 728         results = da.map_blocks(compute_stats, data).compute()
    729 
    730         min_values, max_values, sum_values, ssum_values = results.reshape((-1, 4)).T

~/anaconda3/lib/python3.6/site-packages/dask/base.py in compute(self, **kwargs)
    164         dask.base.compute
    165         """
--> 166         (result,) = compute(self, traverse=False, **kwargs)
    167         return result
    168 

~/anaconda3/lib/python3.6/site-packages/dask/base.py in compute(*args, **kwargs)
    436     postcomputes = [x.__dask_postcompute__() for x in collections]
    437     results = schedule(dsk, keys, **kwargs)
--> 438     return repack([f(r, *a) for r, (f, a) in zip(results, postcomputes)])
    439 
    440 

~/anaconda3/lib/python3.6/site-packages/dask/base.py in <listcomp>(.0)
    436     postcomputes = [x.__dask_postcompute__() for x in collections]
    437     results = schedule(dsk, keys, **kwargs)
--> 438     return repack([f(r, *a) for r, (f, a) in zip(results, postcomputes)])
    439 
    440 

~/anaconda3/lib/python3.6/site-packages/dask/array/core.py in finalize(results)
    953     while isinstance(results2, (tuple, list)):
    954         if len(results2) > 1:
--> 955             return concatenate3(results)
    956         else:
    957             results2 = results2[0]

~/anaconda3/lib/python3.6/site-packages/dask/array/core.py in concatenate3(arrays)
   4313     if not ndim:
   4314         return arrays
-> 4315     chunks = chunks_from_arrays(arrays)
   4316     shape = tuple(map(sum, chunks))
   4317 

~/anaconda3/lib/python3.6/site-packages/dask/array/core.py in chunks_from_arrays(arrays)
   4093 
   4094     while isinstance(arrays, (list, tuple)):
-> 4095         result.append(tuple([shape(deepfirst(a))[dim] for a in arrays]))
   4096         arrays = arrays[0]
   4097         dim += 1

~/anaconda3/lib/python3.6/site-packages/dask/array/core.py in <listcomp>(.0)
   4093 
   4094     while isinstance(arrays, (list, tuple)):
-> 4095         result.append(tuple([shape(deepfirst(a))[dim] for a in arrays]))
   4096         arrays = arrays[0]
   4097         dim += 1

IndexError: tuple index out of range

```

**Comment by e-koch** [comment]
Running more tests now, but I'm finding 7.5 min for `cube.statistics()` (at least until the above error at the end) and ~2 min for CASA's `ia.statistics()` for the same FITS cube.

**Comment by astrofrog** [comment]
@e-koch - could you try again, with bottleneck installed? I've now made sure we use it if possible which should speed things up. Also could you share your cube (could do privately on slack) so that I can try and reproduce the error you are seeing?

**Comment by astrofrog** [comment]
@e-koch @keflavich - for now I've made sure we use the same names of stats as for CASA. I haven't implemented any of the robust ones but presumably that can wait for a separate follow-up PR? The main missing one otherwise is flux and the coordinate-related ones - do we want to add those? (again, maybe a follow-up PR?)

**Comment by astrofrog** [comment]
Not sure what's going on with the Windows build and the open file, but I'll investigate once the PR is otherwise ready

File: spectral_cube/tests/test_dask.py [hunk 1]
```diff
@@ -58,3 +74,39 @@ def test_rechunk(data_adv):
     # note last element is 2 because the chunk size we asked for
     # is larger than cube - this is fine and deliberate in this test
     assert cube_new._data.chunksize == (1, 2, 2)
+
+
+def test_statistics(data_adv):
+    cube = DaskSpectralCube.read(data_adv)
+    stats = cube.statistics()
```
Comment (e-koch) [inline spectral_cube/tests/test_dask.py:81]: @astrofrog I can reproduce (locally) the `IndexError` I commented on by forcing multiple chunks here:
```
    cube = cube.rechunk(chunks=(1, 3, 2))
```

**Comment by astrofrog** [comment]
@e-koch - thanks for the suggestion for reproducing this. I think I've fixed it now, so let me know if it works for you now!

File: spectral_cube/dask_spectral_cube.py [hunk 2]
```diff
@@ -707,6 +707,54 @@ def argmin(self, axis=None, **kwargs):
...
+                             nansum(chunk),
+                             nansum(chunk * chunk)])
+
+        with dask.config.set(**self._scheduler_kwargs):
+            results = da.map_blocks(compute_stats, data.reshape(-1)).compute()
+
+        count_values, min_values, max_values, sum_values, ssum_values = results.reshape((-1, 5)).T
+
+        stats = {'npts': count_values.sum(),
+                 'min': min_values.min() * self._unit,
```
Comment (e-koch) [inline spectral_cube/dask_spectral_cube.py:743]: This should use `self.unit`, I think. When `self.unit` is dimensionless, `self._unit` is `None` and this throws a TypeError

**Comment by e-koch** [comment]
@astrofrog Working with multiple chunks now! I've added a comment above about units (when the cube is dimensionless).

**Comment by astrofrog** [comment]
Thanks! With bottleneck installed, how does the performance compare with CASA?

**Comment by e-koch** [comment]
@astrofrog I'm getting a fairly large variance in run times in both CASA and spectral-cube, but the performance looks comparable!

**Comment by astrofrog** [comment]
Ok so at this point we are probably I/O limited, which is good to know!

**Comment by astrofrog** [comment]
Maybe we can consider merging this and improving the std calculation and adding robust stats in a follow-up.

Just one item for discussion now maybe: do we need an equivalent statistics on the non-dask class?

**Comment by keflavich** [comment]
Yes and yes.  So we should merge this now, and... adding those stats for non-dask is pretty trivial, right?

**Comment by astrofrog** [comment]
@keflavich - yes we can merge this. Adding it to non dask is only trivial if you are ok with whole cubes being loaded into memory!

**Comment by keflavich** [comment]
mean, max, min, sum are all doable as ray/slice iterators with no change.  `std` is implementable via those iterators, but I don't think we have that implementation in hand yet.  We can do the trivial implementation first, but it would be nice to have the iterator version available too.

**Comment by astrofrog** [comment]
Ok I can try and add that in a separate PR?

**Pull request merged**